crates/doc/src/shape/widen.rs

```diff
@@ -2,232 +2,393 @@
             }
-            crate::Node::Number(num) => {
-                self.type_ = self.type_ | types::Set::for_number(&num);
-                false
+            is_first
+        };
+
+        match node.as_node() {
+            Node::Array(items) => {
+                let is_first = apply_type(types::ARRAY);
+                if is_first {
+                    self.array.additional_items = Some(Box::new(Shape::nothing()));
+                }
+                self.array.widen(items, is_first)
             }
-            crate::Node::String(s) => {
-                let is_first_time = !self.type_.overlaps(types::STRING);
-
-                self.type_ = self.type_ | types::STRING;
-
-                // Similar to the nuance around `is_required`, string format
-                // should only "become detected" the very first time. We still
-                // need to run `detect_format` on subsequent strings because
-                // `StringShape::union()` can sometimes widen a string format,
-                // e.g from `integer` -> `number`
-                let format = if is_first_time || self.string.format.is_some() {
-                    Format::detect(s)
+            Node::Bool(_) => apply_type(types::BOOLEAN),
+            Node::Bytes(_) => panic!("not implemented"),
+            Node::Null => apply_type(types::NULL),
+            Node::Float(f) => self.numeric.widen(
+                Number::Float(f),
+                apply_type(if f.fract() != 0.0 {
+                    types::INT_OR_FRAC // Equivalent to "type: number".
                 } else {
-                    None
-                };
-
-                let partial_stringshape = StringShape {
-                    format,
-                    max_length: Some(s.len()),
-                    min_length: s.len(),
-                    ..StringShape::new()
-                };
-
-                if is_first_time {
-                    self.string = partial_stringshape
-                } else {
-                    self.string = StringShape::union(self.string.clone(), partial_stringshape);
+                    types::INTEGER
+                }),
+            ),
+            Node::PosInt(f) => self
+                .numeric
+                .widen(Number::Unsigned(f), apply_type(types::INTEGER)),
+            Node::NegInt(f) => self
+                .numeric
+                .widen(Number::Signed(f), apply_type(types::INTEGER)),
+            Node::Object(fields) => {
+                let is_first = apply_type(types::OBJECT);
+                if is_first {
+                    self.object.additional_properties = Some(Box::new(Shape::nothing()));
                 }
+                self.object.widen::<N>(fields, is_first)
+            }
+            Node::String(s) => self.string.widen(s, apply_type(types::STRING)),
+        }
+    }
 
-                false
+    #[cold]
+    #[inline(never)]
+    fn widen_enum<'n, N>(&mut self, node: &'n N) -> bool
```

Comment:
We should limit the number of enum variants. I could see this one in particular resulting in huge schemas if a field is accidentally defined as an enum, or was previously defined as an enum but then becomes a regular old string field.

Reply:
I think it's okay not too right now.

Enums and tuples would never be produced by schema inference starting from `Shape::nothing()`. Continuously-inferred schemas would never have them.

They _are_ useful for things like `flowctl preview` of a derivation, or suggesting a change of an existing schema given exemplar documents. A particular use case in my mind is SaaS sources that have enum'd schemas for various event types, where a new event type is introduced by the vendor (we've seen a bunch of these).

A commonality is these are non-production, human-in-the-loop usages of widen(), not something we're doing automatically.